source/vir/src/sst_to_air.rs

```diff
@@ -490,6 +499,121 @@ fn one_stmt(stmts: Vec<Stmt>) -> Stmt {
+    if let TypX::Int(range) = &*exp.typ {
+        match range {
+            IntRange::I(_) | IntRange::ISize => {
+                panic!("error: signed integer is not supported for bit-vector reasoning")
+            }
+            _ => (),
+        }
+    };
+}
+
+// convert the sst expression into bv air expression
+fn exp_to_bv_expr(state: &State, exp: &Exp) -> Expr {
+    match &exp.x {
+        ExpX::Const(crate::ast::Constant::Nat(s)) => {
+            if let Some(width) = bitwidth_from_type(&exp.typ) {
+                return Arc::new(ExprX::Const(Constant::BitVec(s.clone(), width)));
+            }
+            panic!("internal error: unable to get width from constant of type {:?}", exp.typ);
+        }
+        ExpX::Var(x) => {
+            return string_var(&suffix_local_unique_id(x));
+        }
+        ExpX::Binary(op, lhs, rhs) => {
+            let bop = match op {
+                BinaryOp::Eq(_) => air::ast::BinaryOp::Eq,
+                BinaryOp::Add => air::ast::BinaryOp::BitAdd,
+                BinaryOp::Sub => air::ast::BinaryOp::BitSub,
+                BinaryOp::Mul => air::ast::BinaryOp::BitMul,
+                BinaryOp::EuclideanDiv => air::ast::BinaryOp::BitUDiv,
+                BinaryOp::EuclideanMod => air::ast::BinaryOp::BitUMod,
+                BinaryOp::Lt => air::ast::BinaryOp::BitULt,
+                BinaryOp::Gt => air::ast::BinaryOp::BitUGt,
+                BinaryOp::Le => air::ast::BinaryOp::BitULe,
+                BinaryOp::Ge => air::ast::BinaryOp::BitUGe,
+                // here the bv operation is translated as it is
+                BinaryOp::BitXor => air::ast::BinaryOp::BitXor,
+                BinaryOp::BitAnd => air::ast::BinaryOp::BitAnd,
+                BinaryOp::BitOr => air::ast::BinaryOp::BitOr,
+                BinaryOp::Shl => air::ast::BinaryOp::Shl,
+                BinaryOp::Shr => air::ast::BinaryOp::LShr,
+                _ => panic!("unhandled bv binary operation {:?}", op),
+            };
+            // disallow signed integer from bitvec reasoning. However, allow that for shift
+            // TODO: sanity check for shift
+            let _ = match op {
+                BinaryOp::Shl | BinaryOp::Shr => (),
+                _ => {
+                    assert_unsigned(&lhs);
+                    assert_unsigned(&rhs);
+                }
+            };
+            let lh = exp_to_bv_expr(state, lhs);
+            let rh = exp_to_bv_expr(state, rhs);
+            return Arc::new(ExprX::Binary(bop, lh, rh));
+        }
+        ExpX::Unary(op, exp) => {
+            // bv type casting by 'as' keyword
+            // convert Clip into concat/extract
+            if let TypX::Int(old_range) = &*exp.typ {
+                if let UnaryOp::Clip(new_range) = op {
+                    match new_range {
+                        IntRange::U(new_n) | IntRange::I(new_n) => match old_range {
+                            IntRange::U(old_n) | IntRange::I(old_n) => {
```

Comment:
Minor style note: you can avoid nesting if/if/match/match by doing nested matching on a pair: `match (&*exp.typ, op) {
                (TypX::Int(IntRange::U(old_n) | IntRange::I(old_n)), UnaryOp::Clip(IntRange::U(new_n) | IntRange::I(new_n))) => ...`

Reply:
It looks much much cleaner this way! Thanks!